firebase/functions/src/post-lab-write.spec.ts

```diff
@@ -0,0 +1,102 @@
+import 'jest';
+import * as sinon from 'sinon';
+import * as admin from 'firebase-admin';
+import * as setupOfflineTests from 'firebase-functions-test';
+
+const test = setupOfflineTests();
+
+test.mockConfig({
+  firebase: {},
+  fb_service_account: {}
+});
+
+describe('Firebase Functions', () => {
+  let postLabWriteFn, hashDirectoryFn, adminInitStub, oldDatabase;
+
+  describe('postLabWrite', () => {
+    beforeEach(() => {
+      adminInitStub = sinon.stub(admin, 'initializeApp');
+      oldDatabase = admin.database;
+      postLabWriteFn = require('./post-lab-write').postLabWrite;
+      hashDirectoryFn = require('./post-lab-write').hashDirectory;
+    });
+
+    afterEach(() => {
+      adminInitStub.restore();
+      admin.database = oldDatabase;
+      test.cleanup();
+    });
+
+    it('shoud update indices and set `cache_hash` for written lab', () => {
+      const labDirectory = [
+        {
+          name: 'main.py',
+          content: ''
+        },
+        {
+          name: 'ml.yaml',
+          content: ''
+        }
+      ];
+
+      const directoryHash = hashDirectoryFn(labDirectory);
+
+      const indicesRefParam = `/idx/lab_executions/some_id`;
+      const labsRefParam = `/labs/some_id/common`;
+
+      const updateLabParam = { cache_hash: directoryHash };
+
+      const databaseStub = sinon.stub();
+      const refStub = sinon.stub();
+      const onceValueStub = sinon.stub();
+      const updateStub = sinon.stub();
+
+      const labSnapshot = {
+        after: {
+          val: () => ({
+            id: 'some_id',
+            user_id: 'some_user_id',
+            directory: labDirectory
+          })
+        }
+      };
+
+      const labExecutionsSnapshot = {
+        val: () => ({ '1': 'first execution', '2': 'section execution' })
+      };
+
+      const contextStub = {
+        params: {
+          id: 'some_id'
+        }
+      };
+
+      const indicesDeltaStub = {
+        '/idx/user_labs/some_user_id/some_id': true,
+        '/idx/user_visible_labs/some_user_id/some_id': true,
+        '/idx/user_visible_executions/some_user_id/1': true,
+        '/idx/user_visible_executions/some_user_id/2': true
+      };
+
+      Object.defineProperty(admin, 'database', {
```

Comment:
Shouldn't this be part of the test setup in `beforeEach`?

Reply:
Can totally be moved there. I guess the reason why it isn't right now is because there might be  tests where it's not needed to stub it.